Dockerfile.gcc: speedup: ccache and Docker cache

christoph-zededa opened the pull request:
to speedup compilation

christoph-zededa committed:
Dockerfile.gcc: enable ccache

to speedup compilation

Signed-off-by: Christoph Ostarek <christoph@zededa.com>

christoph-zededa commented:
will create other PRs once this one is approved

rene commented on Dockerfile.gcc:
```diff
@@ -96,17 +96,14 @@ ENV SOURCE_DATE_EPOCH="${SOURCE_DATE_EPOCH}"
+RUN make -j$(nproc) mrproper
+RUN make O=/kernel-out ${KERNEL_CONFIG}
 RUN --mount=type=cache,target=/root/.cache/ccache,id=kernel-ccache-${TARGETARCH} \
-    ccache -z && \
```
Why are you not resetting the statistics anymore?

christoph-zededa replied:
My fault, I thought it zeros the cache, but it only clears the statistics ...

rene commented on Dockerfile.gcc:
```diff
@@ -96,17 +96,14 @@ ENV SOURCE_DATE_EPOCH="${SOURCE_DATE_EPOCH}"
     && make O=/kernel-out LOCALVERSION="-${LOCALVERSION}" -j$(nproc) modules_install INSTALL_MOD_STRIP=1 \
         INSTALL_MOD_PATH=/tmp/kernel-modules \
-    && ccache -s
+    && ccache -s | tee -a /ccache-stats.txt
```
Why ccache is not enabled? The CC is defined here: https://github.com/lf-edge/eve-kernel/blob/eve-kernel-amd64-v6.1.38-generic/Dockerfile.gcc#L29 with the ccache command included.... at least for arm (cross-compiling) I can ensure it's working...

christoph-zededa replied:
Do you see `ccache` in `ps`? I didn't see it in `execsnoop`.

rene replied:
I didn't check the process, but by applying the following:

```diff
diff --git a/Dockerfile.gcc b/Dockerfile.gcc
index 45286bc2ac9b..9aa68ff823a5 100644
--- a/Dockerfile.gcc
+++ b/Dockerfile.gcc
@@ -98,7 +98,7 @@ RUN ! grep -q -E 'CONFIG_BPF_KPROBE_OVERRIDE=[y|m]' .config || (echo "Cannot set
 
 RUN --mount=type=cache,target=/root/.cache/ccache,id=kernel-ccache-${TARGETARCH} \
     ccache -z && \
-    echo "Building kernel for ${TARGETARCH} with ARCH=${ARCH} and CROSS_COMPILE=${CROSS_COMPILE}" && \
+    echo "Building kernel for ${TARGETARCH} with ARCH=${ARCH} , CC=${CC} and CROSS_COMPILE=${CROSS_COMPILE}" && \
     make -j$(nproc) mrproper \
     && make O=/kernel-out ${KERNEL_CONFIG} \
     && make O=/kernel-out LOCALVERSION="-${LOCALVERSION}" -j$(nproc) prepare \
```

I've got:

```
#17 0.164 Statistics zeroed
#17 0.164 Building kernel for amd64 with ARCH=x86_64 , CC=ccache gcc and CROSS_COMPILE=
```

so I suppose it should work... will check the process as well...

christoph-zededa replied:
@rene ~~an easy~~ another way to test this is do this change:

```
christoph@build-server ~/p/eve-kernel (eve-kernel-amd64-v6.1.68-generic)> git diff
diff --git a/Dockerfile.gcc b/Dockerfile.gcc
index 9e0e780fcdd3..141988cfdcbc 100644
--- a/Dockerfile.gcc
+++ b/Dockerfile.gcc
@@ -96,6 +96,7 @@ ENV SOURCE_DATE_EPOCH="${SOURCE_DATE_EPOCH}"
 # ensure CONFIG_BPF_KPROBE_OVERRIDE is not set
 RUN ! grep -q -E 'CONFIG_BPF_KPROBE_OVERRIDE=[y|m]' .config || (echo "Cannot set this option for security reasons"; false)
 
+ENV CC="asdf ccache gcc"
 RUN --mount=type=cache,target=/root/.cache/ccache,id=kernel-ccache-${TARGETARCH} \
     ccache -z && \
     echo "Building kernel for ${TARGETARCH} with ARCH=${ARCH} and CROSS_COMPILE=${CROSS_COMPILE}" && \
``` 

on `eve-kernel-amd64-v6.1.68-generic` branch and build it with: `docker build -f Dockerfile.gcc .` and then you see it does not fail while building the kernel.

rene replied:
Understood. The problem is not docker, the variable it's exported, but the Makefile from the kernel doesn't consider an environment CC, so as it's solved by this PR, it must be defined at make command line to get priority over the one declared inside the Makefile (which is `$(CROSS_COMPILE)gcc`), and this explains why it was working for arm, because we define the ccache at CROSS_COMPILE and not on CC....

rene replied:
@christoph-zededa , please, don't forget to provide the speedup cache solution (break into another Dockerfile) for arm as well....

christoph-zededa replied:
unfortunately breaking it into two Dockerfiles does only work on some special versions of docker it seems (especially the ones that do not use the more modern buildkit)

rene commented on Dockerfile.gcc:
```diff
@@ -123,9 +120,9 @@ RUN --mount=type=cache,target=/root/.cache/ccache,id=zfs-ccache-${TARGETARCH} \
 
 RUN --mount=type=cache,target=/root/.cache/ccache,id=zfs-ccache-${TARGETARCH} \
-    make -C module -j$(nproc) && \
+    make CC="ccache gcc" -C module -j$(nproc) && \
```
`CC=${CC}` , like you did above....

rucoder replied:
@rene @christoph-zededa it seems ENV is not inherited in some stages, this is expected docker behaviour. I think I lost it while merging stuff.

christoph-zededa committed:
Dockerfile.gcc: enable ccache for HOSTCC

for even more speedup ...

Signed-off-by: Christoph Ostarek <christoph@zededa.com>

rene reviewed (approved):
LGTM

christoph-zededa committed:
Dockerfile.gcc: re-enable docker cache

`COPY . /kernel-src` would always make the cache dirty,
because the sbom files are generated beforehands and have
a new timestamp/content.
So every file except the sbom files are copied extra.
By excluding those files at the beginning, the cache
does not get dirty.

The sbom files are copyied later into the image.

Signed-off-by: Christoph Ostarek <christoph@zededa.com>

rene commented on Dockerfile.gcc:
```diff
@@ -78,10 +78,48 @@ ARG KBUILD_BUILD_TIMESTAMP
+# and we cannot split this into several Dockerfiles as the docker buildkit
+# cannot reference local images
+# To generate this, I used:
+# ls -1 | perl -ne 'chomp; print "COPY $_ /kernel-src/$_\n";'
```
@rucoder , @christoph-zededa , is really `ls -1` enough? Because it's going to list everything, including Dockerfiles, Makefile.eve, etc.... do you think we should include a proper script to list all required files, exclude the ones we want to exclude and generate the COPY lines?

christoph-zededa replied:
I filtered it manually, so `ls -1` is not enough.
I am not sure if it makes sense to create a proper script as the decision whether a file/dir has to copied or not is not so easy to make.
Sure we could have a list of ignored files/dirs, but then instead of editing the output of `ls -1`, one has to edit the list; i.e. I don't see much gain from such a script.

rene replied:
ok, then could you please update the comment `To generate this, I used:` to clarify a bit more, for instance (suggestion), `The list of files can be generated by a command like the following:`, my point is to clarify that only this command might not be enough......

christoph-zededa replied:
I added some text to clarify it: https://github.com/lf-edge/eve-kernel/compare/e621caa498159acd66099403b0e3c81a3763fa62..a56a85e69d6dde9cdd3471614d3fc5b3f0ef0b1e

rene replied:
ok, it's better now. I'm not against these bunch of COPY commands but I'm wondering something..... for nvidia kernel, for example, we have the sources inside another folder, not in the root of the repo... what do you think if we do the same here? You can move the needed files to a kernel-src folder, for example, and then we can just copy this folder with a single COPY command....

christoph-zededa replied:
> what do you think if we do the same here?

I think the problem is that you then cannot rebase anymore on `github.com/torvalds/linux` or do a cherry-pick etc...

christoph-zededa replied:
I thought about some alternatives:
* fix dockerfile-add-scanner
* run dockerfile-add-scanner within this docker container and not separately
* create another repository only with `Makefile.eve`, `Dockerfile.*` which has this repo as a submodule, `repo.sh`, `git clone`, or whatever ...
* put `Makefile.eve`, `Dockerfile.*` into eve's repository under `pkg/kernel`

But I haven't found a definitive good answer.

rene replied:
ok, let's think about it, meanwhile I think we can live with this solution....

rucoder commented:
@rene So according to @deitch we need https://github.com/lf-edge/eve-build-tools/pull/28 to be merged and then bump up the version of buildkit in Makefile.eve to `v0.14.1` so we can use `COPY --exclude`

rucoder commented:
@rene there is confusion here: the PR is submitted agains 6.1.**68** which is wrong. we do not use spdx files for ages. @christoph-zededa asked me to close this one

rucoder commented:
closing this one
